lib/mongo/retryable.rb

```diff
@@ -215,9 +215,13 @@ def write_with_retry(session, write_concern, ending_transaction = false, &block)
         end
         retry_write(e, session, txn_num, &block)
       rescue Error::OperationFailure => e
-        if (session.in_transaction? && !ending_transaction) || !e.write_retryable?
+        if e.code == 20 && e.message.start_with?("Transaction numbers")
```

Comment:
Given that OperationFailure already contains a bunch of logic for figuring out whether a specific error occurred, I think this logic should be moved to OperationFailure in a similar fashion to https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/error/operation_failure.rb#L252.